src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8000,33 +8014,59 @@ private CommonForEachStatementSyntax ParseForEachStatement()
             }
 
             var openParen = this.EatToken(SyntaxKind.OpenParenToken);
+            var variable = ParseExpressionOrDeclaration(ParseTypeMode.Normal, feature: MessageID.IDS_FeatureTuples, permitTupleDesignation: true);
+            var @in = this.EatToken(SyntaxKind.InKeyword, ErrorCode.ERR_InExpected);
+            if (!IsValidForeachVariable(variable))
+            {
+                @in = this.AddError(@in, ErrorCode.ERR_BadForeachDecl);
```

Comment:
Perhaps, but this models the diagnostics we produce currently. I'd prefer to do improvements in a separate PR.